### PR TITLE
feat: use CMD/ENTRYPOINT from source image by default

### LINKED ISSUE #9: Packer overrides CMD and ENTRYPOINT already set in FROM image
_This issue was originally opened by @dimisjim as hashicorp/packer#10311. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

#### Overview of the Issue

A dev familiar with Dockerfile expects Packer to keep the ENTRYPOINT and/or CMD set in the image where the build is based from. Why would Packer reset these, if they are not overridden explicitly in the commit argument of the builder?

#### Reproduction Steps

1. Create a simple nginx image based on the official one.
2. if no CMD is defined in the commit argument of the builder, it defaults to /bin/sh
3. only when `"ENTRYPOINT nginx -g 'daemon off;'"` the image works as expected

### Packer version

1.6.5

### Simplified Packer Buildfile

this is necessary
```
source "docker" "nginx" {
  image   = "nginx"
  commit  = true
  changes = [
    "EXPOSE 80",
    "ENTRYPOINT nginx -g 'daemon off;'"
  ]
}
```
this should be enough:
```
source "docker" "nginx" {
  image   = "nginx"
  commit  = true
  changes = [
    "EXPOSE 80",
  ]
}
```

considering this exists already: 

### Operating system and Environment details

Ubuntu 20.10,
Docker version 19.03.13, build 4484c46

related: #13

### LINKED ISSUE #13: Packer docker builder does not reset ENTRYPOINT properly
_This issue was originally opened by @bertramn as hashicorp/packer#7487. It was migrated here as a result of the [Packer plugin split](https://github.com/hashicorp/packer/issues/8610#issuecomment-770034737). The original body of the issue is below._

<hr>

We are using packer to build a container with a parent that includes an entrypoint. In the packer build we "reset" entrypoint with `ENTRYPOINT []`. Unfortunately the container is being committed with the entrypoint of the parent still in place.

Doing this in a Dockerfile gets rid of the entrypoint from parent:

```Dockerfile
FROM ep-issue:parent
ENTRYPOINT []
CMD []
```

While the same in packer version does not:

```json
{
  "builders": [
    {
      "type": "docker",
      "image": "ep-issue:parent",
      "pull": "false",
      "commit": "true",
      "changes": [
        "ENTRYPOINT []",
        "CMD []"
      ]
    }
  ],
  "post-processors": [
    {
      "type": "docker-tag",
      "repository": "ep-issue",
      "tag": "child2"
    }
  ]
}
```

- Packer version v1.3.5
- Host platform MacOS High Sierra
- Gist of [packer-entrypoint-build.log](https://gist.github.com/bertramn/672485dd92719fae7ed66d20b2f2858e#file-packer-entrypoint-build-log)
- Gist of the [simple script](https://gist.github.com/bertramn/672485dd92719fae7ed66d20b2f2858e#file-reproduce-sh) to reproduce the bug

### LINKED ISSUE #132: simple docker build changing CMD and ENTRYPOINT of resulting image
#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### Overview of the Issue

the resulting image packer produces is changing the CMD and Entrypoint (possibly other things) 

```
build {
  name = "amzn2-ctm-core"

  source "source.docker.dockerhub" {
    image  = "amazonlinux:2"
  }

  post-processors {
    post-processor "docker-tag" {
      tags = [
        "core_${local.timestamp}",
        "core_latest"
      ]
    }
  }
}
```
After building this, comparing the images:

```
# The source image:
$ docker inspect amazonlinux:2 | jq -s ".[0][0].Config.Cmd"
[
  "/bin/bash"
]
$ docker inspect amazonlinux:2 | jq -s ".[0][0].Config.Entrypoint"
null

# the resulting image
$ docker inspect acct1234.dkr.ecr.us-east-1.amazonaws.com/amzn2-ctm:core_20221118194409 | jq -s ".[0][0].Config.Cmd"
null
$ docker inspect acct1234.dkr.ecr.us-east-1.amazonaws.com/amzn2-ctm:core_20221118194409 | jq -s ".[0][0].Config.Entrypoint"
[
  "/bin/sh"
]
```
Why is packer doing this?

### Packer version

1.8.4

### Operating system and Environment details

built on, and sourcing from:
amazonlinux;2

### LINKED ISSUE #158: Packer docker builder reseting ENTRYPOINT and CMD
_This issue was originally opened by @saulfm08 in https://github.com/hashicorp/packer/issues/12529 and has been migrated to this repository. The original issue description is below._

---
We are using packer to build a container with a parent that includes an entrypoint. In the packer build we "reset" entrypoint with ENTRYPOINT []. Unfortunately the container is being committed with the entrypoint of the parent still in place.

Doing this in a Dockerfile gets rid of the entrypoint from parent:
```dockerfile
FROM ep-issue:parent
ENTRYPOINT []
CMD []
```
While the same in packer version does not:
```hcl
source "docker" "wordpress" {
  image  = "wordpress:latest"
  commit = true
}

build {
  name = "my-wp"
  sources = [
    "source.docker.wordpress"
  ]

  post-processors {
    post-processor "docker-tag" {
      repository = "${var.docker_registry}/${var.docker_repository}"
      tags       = ["latest"]
    }
    post-processor "docker-push" {
      ecr_login      = true
      login_server   = var.docker_registry
    }
  }
}
```

```bash
docker image inspect wordpress:latest -f '{{ .ContainerConfig.Cmd }}'
[/bin/sh -c #(nop)  CMD ["apache2-foreground"]]

docker image inspect my-wp:latest -f '{{ .ContainerConfig.Cmd }}'
[]


docker image inspect wordpress:latest -f '{{ .ContainerConfig.Entrypoint }}'
[docker-entrypoint.sh]

docker image inspect my-wp:latest -f '{{ .ContainerConfig.Entrypoint }}'
[/bin/sh]
```

Packer version 1.9.2
Host platform MacOS Ventura

Gist of [packer-entrypoint-build.log](https://gist.github.com/bertramn/672485dd92719fae7ed66d20b2f2858e#file-packer-entrypoint-build-log)
Gist of the [simple script](https://gist.github.com/bertramn/672485dd92719fae7ed66d20b2f2858e#file-reproduce-sh) to reproduce the bug

### PR DESCRIPTION
**Pull request by gardar**
With this change the images built will by default get the CMD and ENTRYPOINT from the source image if not overwritten by a user in the `changes` configuration option.
This replicates the behavior when building images from a Dockerfile.

I found numerous complaints which all seem to be related to the behaviour of the docker builder overwriting the CMD / ENTRYPOINT which this change should solve. (Kind of surprising to see that this hasn't been fixed yet 😵‍)

Closes #158
Closes #132
Closes #13 
Closes #9
Closes https://github.com/hashicorp/packer/issues/4914

### CODE REVIEW

**Comment by gardar** [comment]
ping @nywilken @SwampDragons

**Comment by jkstpierre** [comment]
Can we get this approved and merged? I would love to start using this.

**Review by lbajolet-hashicorp — APPROVED** [review]
Hi @gardar,

Thanks for the patch, this looks good to me!
I'll merge this PR and release the plugin today with that fix in.

**Comment by lbajolet-hashicorp** [comment]
FYI: the new step wasn't `goimport`'d so I took the liberty to run it on the file and add that commit on top of this PR.
I'll wait until tests go green, and I'll merge.

**Comment by gardar** [comment]
> FYI: the new step wasn't `goimport`'d so I took the liberty to run it on the file and add that commit on top of this PR. I'll wait until tests go green, and I'll merge.

Thanks!
Any chance you'll also publish a new release with the fix?

File: builder/docker/step_set_defaults.go
Comment (Spacetown) [inline builder/docker/step_set_defaults.go:35]: This change broke our builds because if both values are defined they are connected. We only defined `ENTRYPOINT` and   got the `CMD` from the base image which then was used as default arguments in our entrypoint.

**Pull request merged**
